File: system_control/libraries/file_loader/test_files/kb_go/postgres/kb_construct/kb_construct_module/ConstructRPCServerTable.go
package kb_construct_module

import (
	"database/sql"
	"encoding/json"
	"fmt"

	"github.com/google/uuid"
	_ "github.com/lib/pq"
)

// ConstructRPCServerTable manages RPC server table operations
type ConstructRPCServerTable struct {
	conn        *sql.DB
	constructKB *ConstructKB
	database    string
	tableName   string
}

// RPCServerFieldResult represents the result of adding an RPC server field
type RPCServerFieldResult struct {
	Status     string                 `json:"status"`
	Message    string                 `json:"message"`
	Properties map[string]interface{} `json:"properties"`
	Data       string                 `json:"data"`
}

// QueueAdjustmentResult represents the result of adjusting a queue
type QueueAdjustmentResult struct {
	Action   string `json:"action"`
	Count    int    `json:"count"`
	NewTotal int    `json:"new_total"`
	Error    string `json:"error,omitempty"`
}

// NewConstructRPCServerTable creates a new instance of ConstructRPCServerTable
func NewConstructRPCServerTable(conn *sql.DB, constructKB *ConstructKB, database string) (*ConstructRPCServerTable, error) {
	crt := &ConstructRPCServerTable{
		conn:        conn,
		constructKB: constructKB,
		database:    database,
		tableName:   database + "_rpc_server",
	}

	if err := crt.setupSchema(); err != nil {
		return nil, fmt.Errorf("error setting up schema: %w", err)
	}

	return crt, nil
}

// setupSchema sets up the database schema
func (crt *ConstructRPCServerTable) setupSchema() error {
	// Drop existing table
	dropQuery := fmt.Sprintf("DROP TABLE IF EXISTS %s CASCADE", crt.tableName)
	if _, err := crt.conn.Exec(dropQuery); err != nil {
		return fmt.Errorf("error dropping table: %w", err)
	}

	// Create the RPC server table
	createTableQuery := fmt.Sprintf(`
		CREATE TABLE %s (
			id SERIAL PRIMARY KEY,
			server_path LTREE NOT NULL,
			
			-- Request information
			request_id UUID NOT NULL DEFAULT gen_random_uuid(),
			rpc_action TEXT NOT NULL DEFAULT 'none',
			request_payload JSONB NOT NULL,
			request_timestamp TIMESTAMPTZ NOT NULL DEFAULT NOW(),
			
			-- Tag to prevent duplicate transactions
			transaction_tag TEXT NOT NULL,
			
			-- Status tracking
			state TEXT NOT NULL DEFAULT 'empty'
				CHECK (state IN ('empty', 'new_job', 'processing')),
			
			-- Additional useful fields
			priority INTEGER NOT NULL DEFAULT 0,
			
			-- New fields as requested
			processing_timestamp TIMESTAMPTZ DEFAULT NULL,
			completed_timestamp TIMESTAMPTZ DEFAULT NULL,
			rpc_client_queue LTREE
		);`, crt.tableName)

	if _, err := crt.conn.Exec(createTableQuery); err != nil {
		return fmt.Errorf("error creating table: %w", err)
	}

	fmt.Println("rpc_server table created.")
	return nil
}

// AddRPCServerField adds a new RPC server field to the knowledge base
func (crt *ConstructRPCServerTable) AddRPCServerField(rpcServerKey string, queueDepth int, description string) (*RPCServerFieldResult, error) {
	properties := map[string]interface{}{
		"queue_depth": queueDepth,
	}
	data := make(map[string]interface{})

	// Add the node to the knowledge base
	if err := crt.constructKB.AddInfoNode("KB_RPC_SERVER_FIELD", rpcServerKey, properties, data, description); err != nil {
		return nil, fmt.Errorf("error adding info node: %w", err)
	}

	fmt.Printf("Added rpc_server field '%s' with properties: %v and data: %v\n", rpcServerKey, properties, data)

	result := &RPCServerFieldResult{
		Status:     "success",
		Message:    fmt.Sprintf("RPC server field '%s' added successfully", rpcServerKey),
		Properties: properties,
		Data:       description,
	}

	return result, nil
}

// RemoveUnspecifiedEntries removes entries from rpc_server_table where server_path is not in the specified list
func (crt *ConstructRPCServerTable) RemoveUnspecifiedEntries(specifiedServerPaths []string) (int, error) {
	if len(specifiedServerPaths) == 0 {
		fmt.Println("Warning: No server_paths specified. No entries will be removed.")
		return 0, nil
	}

	// Filter out empty strings and validate paths
	var validPaths []string
	for _, path := range specifiedServerPaths {
		if path != "" {
			validPaths = append(validPaths, path)
		}
	}

	if len(validPaths) == 0 {
		fmt.Println("Warning: No valid server_paths found after filtering. No entries will be removed.")
		return 0, nil
	}

	fmt.Printf("Processing %d valid server paths\n", len(validPaths))

	// Begin transaction
	tx, err := crt.conn.Begin()
	if err != nil {
		return 0, fmt.Errorf("error beginning transaction: %w", err)
	}
	defer tx.Rollback()

	// Create temporary table
	if _, err := tx.Exec("CREATE TEMP TABLE IF NOT EXISTS valid_server_paths (path text)"); err != nil {
		return 0, fmt.Errorf("error creating temp table: %w", err)
	}

	// Clear the temp table
	if _, err := tx.Exec("DELETE FROM valid_server_paths"); err != nil {
		return 0, fmt.Errorf("error clearing temp table: %w", err)
	}

	// Insert paths in batches
	batchSize := 1000
	insertStmt, err := tx.Prepare("INSERT INTO valid_server_paths VALUES ($1)")
	if err != nil {
		return 0, fmt.Errorf("error preparing insert statement: %w", err)
	}
	defer insertStmt.Close()

	for i := 0; i < len(validPaths); i += batchSize {
		end := i + batchSize
		if end > len(validPaths) {
			end = len(validPaths)
		}

		for j := i; j < end; j++ {
			if _, err := insertStmt.Exec(validPaths[j]); err != nil {
				return 0, fmt.Errorf("error inserting path %s: %w", validPaths[j], err)
			}
		}
	}

	// Set state to empty for remaining records
	updateQuery := fmt.Sprintf(`
		UPDATE %s
		SET state = 'empty'
		WHERE server_path::text IN (
			SELECT path FROM valid_server_paths
		)`, crt.tableName)

	if _, err := tx.Exec(updateQuery); err != nil {
		return 0, fmt.Errorf("error updating state: %w", err)
	}

	// Delete entries not in our temp table
	deleteQuery := fmt.Sprintf(`
		DELETE FROM %s
		WHERE server_path::text NOT IN (
			SELECT path FROM valid_server_paths
		)`, crt.tableName)

	result, err := tx.Exec(deleteQuery)
	if err != nil {
		return 0, fmt.Errorf("error deleting entries: %w", err)
	}

	deletedCount, err := result.RowsAffected()
	if err != nil {
		return 0, fmt.Errorf("error getting rows affected: %w", err)
	}

	// Commit transaction
	if err := tx.Commit(); err != nil {
		return 0, fmt.Errorf("error committing transaction: %w", err)
	}

	// Clean up temp table (best effort)
	crt.conn.Exec("DROP TABLE IF EXISTS valid_server_paths")

	fmt.Printf("Removed %d unspecified entries from %s\n", deletedCount, crt.tableName)
	return int(deletedCount), nil
}

// AdjustQueueLength adjusts the number of records for multiple server paths to match their specified queue lengths
func (crt *ConstructRPCServerTable) AdjustQueueLength(specifiedServerPaths []string, specifiedQueueLengths []int) (map[string]QueueAdjustmentResult, error) {
	results := make(map[string]QueueAdjustmentResult)

	if len(specifiedServerPaths) != len(specifiedQueueLengths) {
		return nil, fmt.Errorf("mismatch between paths and lengths lists")
	}

	// Begin transaction
	tx, err := crt.conn.Begin()
	if err != nil {
		return nil, fmt.Errorf("error beginning transaction: %w", err)
	}
	defer tx.Rollback()

	for i, serverPath := range specifiedServerPaths {
		targetLength := specifiedQueueLengths[i]

		// Get current count
		countQuery := fmt.Sprintf(`
			SELECT COUNT(*) FROM %s 
			WHERE server_path::text = $1`, crt.tableName)

		var currentCount int
		if err := tx.QueryRow(countQuery, serverPath).Scan(&currentCount); err != nil {
			results[serverPath] = QueueAdjustmentResult{Error: err.Error()}
			continue
		}

		// Set state to empty for all records with this server_path
		updateQuery := fmt.Sprintf(`
			UPDATE %s
			SET state = 'empty'
			WHERE server_path::text = $1`, crt.tableName)

		if _, err := tx.Exec(updateQuery, serverPath); err != nil {
			results[serverPath] = QueueAdjustmentResult{Error: err.Error()}
			continue
		}

		if currentCount > targetLength {
			// Remove excess records - oldest first
			deleteQuery := fmt.Sprintf(`
				DELETE FROM %s
				WHERE id IN (
					SELECT id FROM %s
					WHERE server_path::text = $1
					ORDER BY request_timestamp ASC
					LIMIT $2
				)`, crt.tableName, crt.tableName)

			if _, err := tx.Exec(deleteQuery, serverPath, currentCount-targetLength); err != nil {
				results[serverPath] = QueueAdjustmentResult{Error: err.Error()}
				continue
			}

			results[serverPath] = QueueAdjustmentResult{
				Action:   "removed",
				Count:    currentCount - targetLength,
				NewTotal: targetLength,
			}

		} else if currentCount < targetLength {
			// Add placeholder records
			recordsToAdd := targetLength - currentCount

			insertQuery := fmt.Sprintf(`
				INSERT INTO %s (
					server_path, request_payload, transaction_tag, state
				) VALUES (
					$1, $2, $3, 'empty'
				)`, crt.tableName)

			for j := 0; j < recordsToAdd; j++ {
				transactionTag := fmt.Sprintf("placeholder_%s", uuid.New().String())
				if _, err := tx.Exec(insertQuery, serverPath, "{}", transactionTag); err != nil {
					results[serverPath] = QueueAdjustmentResult{Error: err.Error()}
					break
				}
			}

			if results[serverPath].Error == "" {
				results[serverPath] = QueueAdjustmentResult{
					Action:   "added",
					Count:    recordsToAdd,
					NewTotal: targetLength,
				}
			}

		} else {
			results[serverPath] = QueueAdjustmentResult{
				Action:   "unchanged",
				Count:    0,
				NewTotal: currentCount,
			}
		}
	}

	// Commit transaction
	if err := tx.Commit(); err != nil {
		return nil, fmt.Errorf("error committing transaction: %w", err)
	}

	return results, nil
}

// RestoreDefaultValues restores default values for all fields except server_path
func (crt *ConstructRPCServerTable) RestoreDefaultValues() (int, error) {
	updateQuery := fmt.Sprintf(`
		UPDATE %s
		SET 
			request_id = gen_random_uuid(),
			rpc_action = 'none',
			request_payload = '{}'::jsonb,
			request_timestamp = NOW(),
			transaction_tag = CONCAT('reset_', gen_random_uuid()::text),
			state = 'empty',
			priority = 0,
			processing_timestamp = NULL,
			completed_timestamp = NULL,
			rpc_client_queue = NULL
		RETURNING id`, crt.tableName)

	rows, err := crt.conn.Query(updateQuery)
	if err != nil {
		return 0, fmt.Errorf("error updating records: %w", err)
	}
	defer rows.Close()

	updatedCount := 0
	for rows.Next() {
		var id int
		if err := rows.Scan(&id); err != nil {
			return 0, fmt.Errorf("error scanning id: %w", err)
		}
		updatedCount++
	}

	fmt.Printf("Restored default values for %d records\n", updatedCount)
	return updatedCount, nil
}

// CheckInstallation synchronizes the knowledge_base and rpc_server_table
func (crt *ConstructRPCServerTable) CheckInstallation() error {
	// Get specified paths from knowledge_table
	query := fmt.Sprintf(`
		SELECT path, properties FROM %s 
		WHERE label = 'KB_RPC_SERVER_FIELD';`, crt.database)

	rows, err := crt.conn.Query(query)
	if err != nil {
		return fmt.Errorf("error querying knowledge base: %w", err)
	}
	defer rows.Close()

	var paths []string
	var lengths []int

	for rows.Next() {
		var path string
		var propertiesJSON []byte

		if err := rows.Scan(&path, &propertiesJSON); err != nil {
			return fmt.Errorf("error scanning row: %w", err)
		}

		var properties map[string]interface{}
		if err := json.Unmarshal(propertiesJSON, &properties); err != nil {
			return fmt.Errorf("error unmarshaling properties: %w", err)
		}

		paths = append(paths, path)
		
		// Extract queue_depth
		if queueDepth, ok := properties["queue_depth"].(float64); ok {
			lengths = append(lengths, int(queueDepth))
		} else {
			return fmt.Errorf("queue_depth not found or invalid for path %s", path)
		}
	}

	fmt.Printf("paths: %v, lengths: %v\n", paths, lengths)

	// Execute the three operations
	if _, err := crt.RemoveUnspecifiedEntries(paths); err != nil {
		return fmt.Errorf("error removing unspecified entries: %w", err)
	}

	if _, err := crt.AdjustQueueLength(paths, lengths); err != nil {
		return fmt.Errorf("error adjusting queue length: %w", err)
	}

	if _, err := crt.RestoreDefaultValues(); err != nil {
		return fmt.Errorf("error restoring default values: %w", err)
	}

	return nil
}

File: python_programs_and_containers/building_blocks/building_blocks/knowledge_base/construct_kb/construct_rpc_server_table.py
import psycopg2
import json
from psycopg2 import sql
from psycopg2.extensions import adapt
import uuid

class Construct_RPC_Server_Table:
    """
    This class is designed to construct a status table with header
    and info nodes, using a stack-based approach to manage the path. It also
    manages a connection to a PostgreSQL database and sets up the schema.
    """
    def __init__(self, conn, cursor,construct_kb,database):
        self.conn = conn
        self.cursor = cursor
        self.construct_kb = construct_kb
        self.database = database
        self.table_name = self.database + "_rpc_server"
 # Execute the SQL script to set up the schema
        self._setup_schema()

    def _setup_schema(self):
        """
        Sets up the database schema (tables, functions, etc.).
   
        # Use psycopg2.sql module to construct SQL queries safely. This prevents SQL injection.
        # ltree extension needs to be created.
        """
        query = sql.SQL("DROP TABLE IF EXISTS {table_name} CASCADE").format(
            table_name=sql.Identifier(self.table_name)
        )
        self.cursor.execute(query)
        # Create the knowledge_base table
        create_table_script = sql.SQL("""
        CREATE TABLE  {table_name} (
            id SERIAL PRIMARY KEY,
            server_path LTREE NOT NULL,
            
            -- Request information
            request_id UUID NOT NULL DEFAULT gen_random_uuid(),
            rpc_action TEXT NOT NULL DEFAULT 'none',
            request_payload JSONB NOT NULL,
            request_timestamp TIMESTAMPTZ NOT NULL DEFAULT NOW(),
            
            -- Tag to prevent duplicate transactions
            transaction_tag TEXT NOT NULL,
            
            -- Status tracking
            state TEXT NOT NULL DEFAULT 'empty'
                CHECK (state IN ('empty', 'new_job', 'processing')),
            
            -- Additional useful fields
            priority INTEGER NOT NULL DEFAULT 0,
            
            -- New fields as requested
            processing_timestamp TIMESTAMPTZ DEFAULT NULL,
            completed_timestamp TIMESTAMPTZ DEFAULT NULL,
            rpc_client_queue LTREE
        );
        """).format(table_name=sql.Identifier(self.table_name))


        self.cursor.execute(create_table_script)
        self.conn.commit()  # Commit the changes
        print("rpc_server table created.")

    def add_rpc_server_field(self, rpc_server_key,queue_depth, description):
        """
        Add a new status field to the knowledge base
        
        Args:
            rpc_server_key (str): The key/name of the status field
            queue_depth (int): The length of the rpc_server
            description (str): The description of the rpc_server
            
        Raises:
            TypeError: If status_key is not a string or properties is not a dictionary
        """
        if not isinstance(rpc_server_key, str):
            raise TypeError("rpc_server_key must be a string")
        
        if not isinstance(queue_depth, int):
            raise TypeError("queue_depth must be an integer")
        if not isinstance(description, str):
            raise TypeError("description must be a string")
        properties = {'queue_depth': queue_depth}
 
        data = {}
        
        # Add the node to the knowledge base
        self.construct_kb.add_info_node("KB_RPC_SERVER_FIELD", rpc_server_key, properties, data,description)

        print(f"Added rpc_server field '{rpc_server_key}' with properties: {properties} and data: {data}")
        
        return {
            "status": "success",
            "message": f"RPC server field '{rpc_server_key}' added successfully",
            "properties": properties,
            "data": description
        }
            
    def remove_unspecified_entries(self, specified_server_paths):
        """
        Remove entries from rpc_server_table where the server_path is not in the specified list,
        handling large lists of paths efficiently.
        
        Args:
            specified_server_paths (list): List of valid server_paths to keep
            
        Returns:
            int: Number of deleted records
        """
        try:
            if not specified_server_paths:
                print("Warning: No server_paths specified. No entries will be removed.")
                return 0
            
            # Filter out None values and convert all paths to strings
            valid_paths = []
            for path in specified_server_paths:
                if path is not None:
                    valid_paths.append(str(path))
            
            if not valid_paths:
                print("Warning: No valid server_paths found after filtering. No entries will be removed.")
                return 0
            
            print(f"Processing {len(valid_paths)} valid server paths")
            
            # Create a temporary table to store valid paths for efficient processing
            # Don't use explicit transaction control to avoid set_session conflicts
            self.cursor.execute("""
                CREATE TEMP TABLE IF NOT EXISTS valid_server_paths (path text)
            """)
            
            # Clear the table in case it already exists
            self.cursor.execute("DELETE FROM valid_server_paths")
            
            # Insert paths in batches to avoid parameter limits
            batch_size = 1000
            for i in range(0, len(valid_paths), batch_size):
                batch = valid_paths[i:i+batch_size]
                args = [(path,) for path in batch]
                try:
                    self.cursor.executemany("""
                        INSERT INTO valid_server_paths VALUES (%s)
                    """, args)
                except Exception as batch_error:
                    print(f"Error inserting batch {i//batch_size + 1}: {batch_error}")
                    print(f"Problematic batch: {batch}")
                    raise
            
            # Set state to empty for remaining records before deleting unspecified ones
            update_query = sql.SQL("""
                UPDATE {table_name}
                SET state = 'empty'
                WHERE server_path::text IN (
                    SELECT path FROM valid_server_paths
                )
            """).format(table_name=sql.Identifier(self.table_name))
            
            self.cursor.execute(update_query)
            
            # Delete entries where server_path is not in our temp table
            delete_query = sql.SQL("""
                DELETE FROM {table_name}
                WHERE server_path::text NOT IN (
                    SELECT path FROM valid_server_paths
                )
            """).format(table_name=sql.Identifier(self.table_name))
            
            self.cursor.execute(delete_query)
            deleted_count = self.cursor.rowcount
            
            # Commit the transaction
            self.conn.commit()
            
            # Clean up the temporary table
            try:
                self.cursor.execute("DROP TABLE IF EXISTS valid_server_paths")
            except Exception as cleanup_error:
                print(f"Warning: Could not clean up temporary table: {cleanup_error}")
            
            print(f"Removed {deleted_count} unspecified entries from {self.table_name}")
            return deleted_count
            
        except Exception as e:
            print(f"Error in remove_unspecified_entries: {e}")
            # Roll back in case of error - but don't force autocommit changes
            try:
                if hasattr(self, 'conn') and self.conn:
                    self.conn.rollback()
            except Exception as rollback_error:
                print(f"Warning: Could not rollback transaction: {rollback_error}")
            raise Exception(f"Error in remove_unspecified_entries: {e}")
        import uuid

    def adjust_queue_length(self, specified_server_paths, specified_queue_lengths):
        """
        Adjust the number of records for multiple server paths to match their specified queue lengths.
        
        Args:
            specified_server_paths (list): List of server paths to adjust
            specified_queue_lengths (list): List of desired queue lengths corresponding to each server path
            
        Returns:
            dict: A dictionary with server paths as keys and operation results as values
        """
        results = {}
        
        try:
            if len(specified_server_paths) != len(specified_queue_lengths):
                raise ValueError("Mismatch between paths and lengths lists")
                
            for i, server_path in enumerate(specified_server_paths):
                try:
                    target_length = int(specified_queue_lengths[i])
                    
                    # Get current count
                    count_query = sql.SQL("""
                        SELECT COUNT(*) FROM {table_name} 
                        WHERE server_path::text = %s
                    """).format(table_name=sql.Identifier(self.table_name))
                    
                    self.cursor.execute(count_query, (server_path,))
                    current_count = self.cursor.fetchone()[0]
                    
                    # Set state to empty for all records with this server_path
                    update_query = sql.SQL("""
                        UPDATE {table_name}
                        SET state = 'empty'
                        WHERE server_path::text = %s
                    """).format(table_name=sql.Identifier(self.table_name))
                    
                    self.cursor.execute(update_query, (server_path,))
                    
                    if current_count > target_length:
                        # Need to remove excess records - remove oldest first
                        delete_query = sql.SQL("""
                            DELETE FROM {table_name}
                            WHERE id IN (
                                SELECT id FROM {table_name}
                                WHERE server_path::text = %s
                                ORDER BY request_timestamp ASC
                                LIMIT %s
                            )
                        """).format(table_name=sql.Identifier(self.table_name))
                        
                        self.cursor.execute(delete_query, (server_path, current_count - target_length))
                        
                        results[server_path] = {
                            'action': 'removed',
                            'count': current_count - target_length,
                            'new_total': target_length
                        }
                        
                    elif current_count < target_length:
                        # Need to add placeholder records
                        records_to_add = target_length - current_count
                        
                        insert_query = sql.SQL("""
                            INSERT INTO {table_name} (
                                server_path, request_payload, transaction_tag, state
                            ) VALUES (
                                %s, %s, %s, 'empty'
                            )
                        """).format(table_name=sql.Identifier(self.table_name))
                        
                        for _ in range(records_to_add):
                            self.cursor.execute(insert_query, (
                                server_path,  # Use server_path as server_path for placeholders
                                '{}',         # Empty JSON object
                                f"placeholder_{uuid.uuid4()}"  # Unique transaction tag
                            ))
                        
                        results[server_path] = {
                            'action': 'added',
                            'count': records_to_add,
                            'new_total': target_length
                        }
                        
                    else:
                        results[server_path] = {
                            'action': 'unchanged',
                            'count': 0,
                            'new_total': current_count
                        }
                        
                except Exception as path_error:
                    print(f"Error adjusting queue for path {server_path}: {path_error}")
                    results[server_path] = {'error': str(path_error)}
            
            # Commit all changes
            self.conn.commit()
            return results
            
        except Exception as e:
            print(f"Error in adjust_queue_length: {e}")
            if hasattr(self, 'conn') and self.conn:
                self.conn.rollback()
            raise Exception(f"Error in adjust_queue_length: {e}")
        
        
    def restore_default_values(self):
        """
        Restore default values for all fields in rpc_server_table except for server_path.
        
        This method will:
        1. Generate a unique UUID for request_id for each record
        2. Set rpc_action to 'none'
        3. Set request_payload to an empty JSON object
        4. Set request_timestamp to current time
        5. Generate a new transaction_tag
        6. Set state to 'empty'
        7. Set priority to 0
        8. Clear processing_timestamp (set to NULL)
        9. Clear completed_timestamp (set to NULL)
        10. Clear rpc_client_queue (set to NULL)
        
        Returns:
            int: Number of records updated
        """
        # Update all records with default values while preserving server_path
        update_query = sql.SQL("""
            UPDATE {table_name}
            SET 
                request_id = gen_random_uuid(),
                rpc_action = 'none',
                request_payload = '{{}}'::jsonb,
                request_timestamp = NOW(),
                transaction_tag = CONCAT('reset_', gen_random_uuid()::text),
                state = 'empty',
                priority = 0,
                processing_timestamp = NULL,
                completed_timestamp = NULL,
                rpc_client_queue = NULL
            
            RETURNING id
        """).format(table_name=sql.Identifier(self.table_name))
        
        self.cursor.execute(update_query)
        
        # Get count of updated records using RETURNING clause
        updated_count = len(self.cursor.fetchall())
        
        print(f"Restored default values for {updated_count} records")
        return updated_count
        
    def check_installation(self):     

        
            # Get specified paths (paths with label "KB_RPC_SERVER_FIELD") from knowledge_table
            query = sql.SQL("""
                SELECT path, properties FROM {table_name} 
                WHERE label = 'KB_RPC_SERVER_FIELD';
            """).format(table_name=sql.Identifier(self.database))
            
            self.cursor.execute(query)
            specified_paths_data = self.cursor.fetchall()
            
            paths = []
            lengths = []
            for row in specified_paths_data:
                paths.append(row[0])
                properties = row[1]
                lengths.append(properties['queue_depth'])
            print(f"paths: {paths}", f"lengths: {lengths}")

            self.remove_unspecified_entries(paths)
            self.adjust_queue_length(paths, lengths)
            self.restore_default_values()
            
  